pkg/abcupdater/metrics.go

```diff
@@ -0,0 +1,226 @@
+	lookuper := info.Lookuper
+	if lookuper == nil {
+		lookuper = envconfig.OsLookuper()
+	}
+
+	optOutSettings, err := loadOptOutSettings(ctx, lookuper, info.AppID)
+	if err != nil {
+		return fmt.Errorf("failed to load opt out settings: %w", err)
+	}
+
+	if optOutSettings.NoMetrics {
+		return nil
+	}
+
+	generateNewID := true
+	storedID, err := loadInstallID(info)
+	if err == nil && storedID != nil {
+		oneDayAgo := time.Now().Add(-24 * time.Hour)
+		// Defensively check for case ID Created is in future.
+		generateNewID = oneDayAgo.Unix() >= storedID.IDCreatedTimestamp ||
+			time.Now().Unix() < storedID.IDCreatedTimestamp
+	}
+	var installID string
+	if generateNewID {
+		installUUID, err := uuid.NewRandom()
+		if err != nil {
+			return fmt.Errorf("could not generate id for metrics: %w", err)
+		}
+		installID = installUUID.String()
+		err = storeInstallID(info, &InstallIDData{
+			IDCreatedTimestamp: time.Now().Unix(),
+			InstallID:          installID,
+		})
+		if err != nil {
+			logging.FromContext(ctx).DebugContext(ctx, "error storing installID", "error", err.Error())
+		}
+	} else {
+		installID = storedID.InstallID
+	}
+
+	var c metricsConfig
+	if err := envconfig.ProcessWith(ctx, &envconfig.Config{
+		Target:   &c,
+		Lookuper: lookuper,
+	}); err != nil {
+		return fmt.Errorf("failed to process envconfig: %w", err)
+	}
+
+	// Use ParseRequestURI over Parse because Parse validation is more loose and will accept
+	// things such as relative paths without a host.
+	if _, err := url.ParseRequestURI(c.ServerURL); err != nil {
+		return fmt.Errorf("failed to parse server url: %w", err)
+	}
+
+	client := &http.Client{}
+
+	buf := bytes.Buffer{}
+	if err := json.NewEncoder(&buf).Encode(SendMetricRequest{
+		AppID:     info.AppID,
+		Version:   info.Version,
+		Metrics:   info.Metrics,
+		InstallID: installID,
+	}); err != nil {
+		return fmt.Errorf("failed to marshal metric json: %w", err)
+	}
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodPost, fmt.Sprintf(c.ServerURL+"/sendMetrics"), &buf)
+	if err != nil {
+		return fmt.Errorf("failed to create request: %w", err)
+	}
+
+	resp, err := client.Do(req)
+	if err != nil {
+		return fmt.Errorf("failed to make request: %w", err)
+	}
+	defer resp.Body.Close()
+
+	// Future releases may be more strict.
+	if resp.StatusCode >= 300 || resp.StatusCode <= 199 {
+		b, err := io.ReadAll(io.LimitReader(resp.Body, maxErrorResponseBytes))
+		if err != nil {
+			return fmt.Errorf("received %d response, unable to read response body", resp.StatusCode)
+		}
+
+		return fmt.Errorf("received %d response: %s", resp.StatusCode, string(b))
+	}
+
+	// For now, ignore response body for happy responses.
+	// Future versions may parse warnings for debug logging.
+	return nil
+}
+
+// A per-application install id is randomly generated. It gets rotated every
+// 24 hours. This is to avoid a single client polluting metrics.
+func loadInstallID(c *MetricsInfo) (*InstallIDData, error) {
```

Comment:
I'm a little confused... 

1. Why would we want a different install id per client? 
2. Why can't this just be a plaintext file on disk that contains a random string (os/file.read), and we can use the modtime of the file to determine if there's a need to rotate or expire the identifier?
3. There's a lot of "override this thing for testing", which is often a design smell. It seems like the `MetricsClient` should load or create a stable ID on boot, and then it's string from that point forward in all the other APIs.

Reply:
1. To avoid a single misbehaving client to pollute any metrics values. Also so we can have an idea on whether a metric is from a CI/CD instance (we will only see id once) or a more long lived process.

2. Expiration is no longer used. I am keeping the install time for now, as we may find it useful later, though its not exposed to client and isn't sent to server.

3. I think I'll do that.

Reply:
What is a "client" in this case, and how is the install id different from the UUID?

Reply:
Client is anyone calling our metrics API. The threat I'm thinking about is an accidental infinite loop skewing all our data. It would be difficult to defend against a motivated person trying to mess with our metrics, as we are not requiring users to authenticate.

Install ID is the UUID. I'm actually changing it to be just a base64 encoded 64 bit random ID. I wanted to reuse the `localstore` library used for metrics, which already handles all the json stuff and nicely encapsulates the file IO.

Reply:
I guess I'm not sure I understand the difference between the install id and the client id then? Wouldn't we just block/ignore a spammy install id?

Reply:
Yes, that's the idea. If I used the term client ID, I meant install ID. I only have one ID.

Reply:
Oh then that's fine. I thought we had one id per app per client.

Reply:
Let me make sure we are on the same page.

Install ID is per application that is using the updater metrics library.

`abc` for example on a single computer would always have the same ID, say `FOOBAR`
`jvs` or some other program on that same computer would have its own ID, say `BAZBUZZ`

The idea is that I want to somewhat isolate the dependency of applications on a particular file format or location. If someone updates `abc` but not `jvs`, you could have thrashing between different versions of the metrics library if we ever decide to change file format.

Reply:
Hmm - I think we have two string values, of which I'll use obviously-wrong names so we can focus on the definintion

- "BANANA" - identifier of an app (e.g. "jvsctl" or "abc") which is constant for the app. All instances of the app use the same identifier.

- "PANCAKE" - per-machine identifier which is a randomly-generated collection of bytes used to uniquely identify the machine (without other PII).

When we check for updates, we check via "BANANA", and versions/responses are scoped per-BANANA. When we send metrics, those metrics include BANANA and PANCAKE.

Reply:
Almost correct.

So you have your apps, lets call them BANANA and MANGO.

Updates check for BANANA sends BANANA, and for MANGO sends MANGO.

Current design has a per-machine-per-app identifier. So BANANA will send metrics with the per-machine identifier PANCAKE, but MANGO would instead have its own per-machine identifier FRENCH_TOAST

Reply:
What's the reason for that? It seems like we'd _want_ to correlate that someone is a user of both TOOL_A and TOOL_B?

Reply:
Yeah, that would be handy. I'm running it by privacy eng to ensure doesn't raise any concerns as it differs from design doc. Moving to issue since I want to get this merged: https://github.com/abcxyz/abc-updater/issues/57